[fix] make /stats more CSP compliant

return42 opened the pull request:
## What does this PR do?

Make /stats more CSP compliant, for details read commit messages.

## Why is this change important?

Reported-by: #57

## How to test this PR locally?

Test locally (see below), before we merge I will test this PR with CSP _activated_ on my [public instance](https://darmarit.org/searx/stats).

Aktivate checker ..

```diff
modified   searx/settings.yml
@@ -106,7 +106,7 @@ outgoing: # communication with search engines
 
 checker:
     # disable checker when in debug mode
-    off_when_debug: True
+    off_when_debug: False
 
     # scheduling: interval or int
     # use "scheduling: False" to disable scheduling
@@ -114,9 +114,9 @@ checker:
     # * uncomment "scheduling" section
     # * add "cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1" to your uwsgi.ini
 
-    # scheduling:
-    #    start_after: [300, 1800]  # delay to start the first run of the checker
-    #    every: [86400, 90000]  # how often the checker runs
+    scheduling:
+        start_after: [10, 20]  # delay to start the first run of the checker
+        every: [1000, 2000]  # how often the checker runs
 
     # additional tests: only for the YAML anchors (see the engines section)
```

Start instance ..

    make run 

Doe some queries, wait until checker has finished .. and after visit /stats page

return42 committed:
[fix] make /stats more CSP compliant - bar graph

Replace bar graph's *styles* (see below) by CSP compliant implementation in
``searx/static/themes/__common__/less/stats.less`` ::

    ./simple/stats.html:49: <span style="width: calc(max(2px, 100%*{{ (engine_stat.result_count / engine_stats.max_result_count )|round(3) }}))" class="stacked-bar-chart-serie1"></span>{{- "" -}}
    ./simple/stats.html:57: <span style="width: calc(max(2px, 100%*{{ (engine_stat.http / engine_stats.max_time )|round(3) }}))" class="stacked-bar-chart-serie1"></span>{{- "" -}}
    ./simple/stats.html:58: <span style="width: calc(100%*{{ engine_stat.processing / engine_stats.max_time |round(3) }})" class="stacked-bar-chart-serie2"></span>{{- "" -}}
    ./oscar/stats.html:50:  <span style="width: calc(max(2px, 100%*{{ (engine_stat.result_count / engine_stats.max_result_count )|round(3) }}))" class="stacked-bar-chart-serie1"></span>{{- "" -}}
    ./oscar/stats.html:58:  <span style="width: calc(max(2px, 100%*{{ (engine_stat.http / engine_stats.max_time )|round(3) }}))" class="stacked-bar-chart-serie1"></span>{{- "" -}}
    ./oscar/stats.html:59:  <span style="width: calc(100%*{{ engine_stat.processing / engine_stats.max_time |round(3) }})" class="stacked-bar-chart-serie2"></span>{{- "" -}}

Reported-by: https://github.com/searxng/searxng/issues/57
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] make /stats more CSP compliant - simple theme

Replace simple theme's *styles* (see below) by CSP compliant implementation in
``searx/static/themes/simple/less/stats.less`` ::

    ./simple/stats.html:26:  <table style="max-width: 1280px; margin: 0 auto 0 0;">
    ./simple/stats.html:28:  <th scope="col" style="width:20rem;">{{ th_sort('name', _("Engine name")) }}</th>
    ./simple/stats.html:29:  <th scope="col" style="width:7rem; text-align: right;">{{ th_sort('score', _('Scores')) }}</th>
    ./simple/stats.html:32:  <th scope="col" style="text-align: right;">{{ th_sort('reliability', _('Reliability')) }}</th>
    ./simple/stats.html:37:  <td style="text-align: right;">
    ./simple/stats.html:90:  <td style="text-align: right;"> {{ engine_reliabilities.get(engine_stat.name, {}).get('reliablity') }}</td>
    ./simple/stats.html:106: <table style="max-width: 1280px; margin: 1rem; border: 1px solid gray;">
    ./simple/stats.html:107: <tbody style="padding-top: 1rem;">
    ./simple/stats.html:110: <th scope="row" style="width: 10rem">{{ _('Exception') }}</th><td>{{ error.exception_classname }}</td>
    ./simple/stats.html:112: <th scope="row" style="width: 10rem">{{ _('Message') }}</th><td>{{ error.log_message }}</td>
    ./simple/stats.html:114: <th scope="row" style="width: 10rem">{{ _('Percentage') }}</th><td style="width: 10rem">{{ error.percentage }}</td>
    ./simple/stats.html:119: <span style="border-right: 1px solid gray; padding: 0 1rem 0 0; margin: 0 0 0 0.5rem;">{{ param }}</span>
    ./simple/stats.html:136: <th scope="col" style="width: 10rem">{{ _('Failed test') }}</th>

Reported-by: https://github.com/searxng/searxng/issues/57
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] make /stats more CSP compliant - oscar theme

Replace oscar theme's *styles* (see below) by CSP compliant implementation in
``searx/static/themes/__common__/less/stats.less`` ::

    ./oscar/stats.html:29:  <th scope="col" style="width:20rem;">{{ th_sort('name', _("Engine name")) }}</th>
    ./oscar/stats.html:30:  <th scope="col" style="width:7rem; text-align: right;">{{ th_sort('score', _('Scores')) }}</th>
    ./oscar/stats.html:33:  <th scope="col" style="text-align: right;">{{ th_sort('reliability', _('Reliability')) }}</th>
    ./oscar/stats.html:38:  <td style="text-align: right;">
    ./oscar/stats.html:91:  <td style="text-align: right;"> {{ engine_reliabilities.get(engine_stat.name, {}).get('reliablity') }}</td>
    ./oscar/stats.html:109: <tbody style="padding-top: 1rem;">
    ./oscar/stats.html:112: <th scope="row" style="width: 10rem">{{ _('Exception') }}</th><td>{{ error.exception_classname }}</td>
    ./oscar/stats.html:114: <th scope="row" style="width: 10rem">{{ _('Message') }}</th><td>{{ error.log_message }}</td>
    ./oscar/stats.html:116: <th scope="row" style="width: 10rem">{{ _('Percentage') }}</th><td style="width: 10rem">{{ error.percentage }}</td>
    ./oscar/stats.html:121: <span style="border-right: 1px solid gray; padding: 0 1rem 0 0; margin: 0 0 0 0.5rem;">{{ param }}</span>
    ./oscar/stats.html:138: <th scope="col" style="width: 10rem">{{ _('Failed test') }}</th>

Reported-by: https://github.com/searxng/searxng/issues/57
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] make /stats more CSP compliant - github issue form

Hide textarea from github issue form::

    ./__common__/new_issue.html:6:    <textarea name="body" style="display: none;">{{- '' -}}

BTW: fix indentation.

Reported-by: https://github.com/searxng/searxng/issues/57
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on searx/static/themes/__common__/less/new_issue.less:
```diff
@@ -9,6 +9,11 @@
+
+.issue-hide {
+  display: none;
+}
```
Related to the github new issue template not to stats.less

dalf commented on searx/static/themes/__common__/less/stats.less:
```diff
@@ -1,60 +1,54 @@
 
-    table .engine-scores {
-	width: 7rem;
+    .engine-score {
```
see 
* https://github.com/return42/searxng/blob/1204e4f07e932029d66b4674a3474918228459a6/searx/templates/oscar/stats.html#L27
* https://github.com/return42/searxng/blob/1204e4f07e932029d66b4674a3474918228459a6/searx/templates/simple/stats.html#L26

dalf commented on searx/static/themes/__common__/less/stats.less:
```diff
@@ -1,60 +1,54 @@
-.issue-hide{
-    display: none;
-}
-
```
The file is untabify.

Is it possible not to add tab ?
I guess it is more complicated than `(setq-default indent-tabs-mode nil)` ?

return42 replied:
Sorry, I'm sloppy and use the emacs defaults for new created files / For better tab handling I set `indent-tabs-mode` in 66c84827.

dalf commented on searx/static/themes/__common__/less/stats.less:
```diff
@@ -1,60 +1,54 @@
 .bar-chart-graph {
-    width: 300px;
+    width: calc(100% - 5rem);
+    display: inline-block;
```
From line 44 to line 51.

There is a slight visual change with this PR:
![image](https://user-images.githubusercontent.com/1594191/118454524-c1881e00-b6f8-11eb-86b2-3fd4f812aa7b.png)

The layout in the master branch is:
![image](https://user-images.githubusercontent.com/1594191/118454584-d2d12a80-b6f8-11eb-970a-66997e28096f.png)

return42 replied:
I haven't found the solution,  thanks a lot for this fix.

dalf commented on searx/static/themes/simple/package.json:
```diff
@@ -4,7 +4,7 @@
     "grunt-contrib-cssmin": "^2.2.1",
     "grunt-contrib-jshint": "~1.1.0",
-    "grunt-contrib-less": "^1.4.1",
+    "grunt-contrib-less": "~3.0.0",
```
Upgraded otherwise:
`width: calc(100% - 5rem);` becomes `width: 95%` once compiled by less version 1.4.1.

dalf commented on searx/templates/oscar/stats.html:
```diff
@@ -27,15 +27,15 @@ <h1>{{ _('Engine stats') }}{% if selected_engine_name %} - {{ selected_engine_na
                         <tr>
                             <td class="engine-name"><a href="{{ url_for('stats', engine=engine_stat.name|e) }}">{{ engine_stat.name }}</a></td>
-                            <td class="engine-scores">
+                            <td class="engine-score">
```
`engine-score` -> `engine-scores` same in the .less files above.
It exists before this PR.

dalf commented on searx/static/themes/simple/gruntfile.js:
```diff
@@ -68,11 +68,12 @@ module.exports = function(grunt) {
+          sourceMap: true,
+          sourceMapURL: (name) => { const s = name.split('/'); return s[s.length - 1] + '.map';},
+          outputSourceFiles: false,
+          sourceMapRootpath: '../',
```
Copy / paste from searx/static/themes/oscar/gruntfile.js

return42 committed:
[less] update grunt-contrib-less v3.3.0

Upgraded [v3.3.0] otherwise::

`  width: calc(100% - 5rem);`

becomes `width: 95%` once compiled by less version 1.4.1.

[v3.3.0] https://github.com/gruntjs/grunt-contrib-less/releases/tag/v3.0.0

Suggested-by: @dalf in commit 1204e4f0
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] simple theme - generate missing sourceMap file

C&P from searx/static/themes/oscar/gruntfile.js

Suggested-by: @dalf in commit 1204e4f0
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] bar graph and rename CSS class engine-scores -> engine-score

- drop #main_stats selector in stats.less
- 'engine-score' exists before this PR.
- untabify searx/static/themes/__common__/less/stats.less

for details see comment at: https://github.com/searxng/searxng/pull/81/files/d93bec7638908cae64530727d4e1539ae10bd144..1204e4f07e932029d66b4674a3474918228459a6#r633571496

Suggested-by: @dalf in commit 1204e4f0
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[static] make /stats more CSP compliant- make themes.all

Based on commits

- 0507e185 [fix] bar graph and rename CSS class engine-scores -> engine-score
- 3e9ad7ae [fix] make /stats more CSP compliant - github issue form
- 34859d0e [fix] make /stats more CSP compliant - oscar theme
- 0a6c4884 [fix] make /stats more CSP compliant - simple theme
- cdfb4b7f [fix] make /stats more CSP compliant - bar graph
- 965817f2 [fix] simple theme - generate missing sourceMap file

this patch is generated by::

     make themes.all

Reported-by: https://github.com/searxng/searxng/issues/57
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[emacs] don't use tabs, except in makefile-gmake-mode

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>
